### DISCUSSION TITLE
Frequency-based keymap modiƒew

### DISCUSSION THREAD

**Discussion by eugenesvk** [helix-editor/helix#6665] [Show and tell]
[full info here](https://eugenesvk.github.io/kbdHelix/modifew/)

I'd like to show&tell about the __modiƒew__ keymap that tries to break the chains ⛓ of dubious keycap-based mnemonics like __f__ for `find` (or is it `forward`?) and instead maps the most __frequent commands__ to the __best keys__, so <kbd>f</kbd>=👆🏻 `moves forward by a word` instead<br>It also attempts to convert most of the commands requiring modifiers into a chainable key sequence, hence its name: "few modifiers based on frequency(ƒ)" or __modiƒew__

Some of the design improvements:
  - ⌂ __Home__ keys are __safe__: all 8 of the home keys are movement commands, so to modify a buffer your fingers need to leave __⌂__!
  - __Insert mode__ has a copy of most of the Normal mode commands (from <kbd>key</kbd> and <kbd>⇧</kbd><kbd>key</kbd>) on the sames keys, but with an extra <kbd>⌥</kbd> — no need to switch modes for just a single command!
  - __Space mode__ converts most of the commands requiring modifiers into a chainable key sequence with a "leader" key <kbd>Space</kbd> as the easiest to press. Its main layer mostly consists of commands that were assigned to <kbd>⎈</kbd> or <kbd>⇧</kbd> in the Normal mode, though via additional submodes any other command can also be chained (and if this [feature request](https://github.com/helix-editor/helix/issues/1499) is implemented, this approach could also work nicely with repeatable commands that would not close the menu dialog command list upon execution)
  - Differentiate between __repeatable__ and __single-press__ __paired__ keybinds (left/right, back/forward, earlier/later etc.)
    + __Repeatable__ keybinds should be on __separate__ __adjacent__ keys (just like <kbd>◀</kbd><kbd>▶</kbd>) instead of being on the same key with the second operation behind a <kbd>⇧</kbd>’ed status, so char/word/line-based movements and in-/decrement are together rather than being on separate rows by default. This greatly simplifies going back and forth:
    + __Single-press__ paired keybinds can remain with the opposite direction <kbd>⇧</kbd>’ed, e.g. move to the beginning/end of a line <kbd lyt=q>a</kbd><span class="keyMove">⭰</span> <kbd>⇧</kbd><kbd lyt=q>a</kbd><span class="keyMove">⭲</span>
  - Bonus: the `Cut`/`Copy`/`Paste` commands' frequencies seem to allow having them together under 💍🖕👆 left hand bottom row with the common <kbd lyt=q>x</kbd><kbd lyt=q>c</kbd><kbd lyt=q>v</kbd>, though `Undo` is too frequent to be left at a 🤙<kbd lyt=q>z</kbd> (however, with <kbd>⎈</kbd> they all maintain the	<kbd lyt=q>z</kbd><kbd lyt=q>x</kbd><kbd lyt=q>c</kbd><kbd lyt=q>v</kbd> compatibility)

Now, to the layout itself: while it might be intimidating at first with all the icon jazz, it's a great reference point as you get used to it as you can get all the commands per mode in a single cheat sheet (the config file has the same icons to make the lookup a bit easier)

![modifew Normal Mode](https://raw.githubusercontent.com/eugenesvk/kbdHelix/modifew/img/helix-keymap-modifew-m1NOR.png)

The full description of the design principles with all the modes (images and html layouts) as well as the config itself (and a tip on how to make updating it less of a chore) at this page https://eugenesvk.github.io/kbdHelix/

**Comment by well1791** [comment]
finally someone decided to do what actually makes sense!

**Comment by willfaught** [comment]
I read this post and your page, but I don't understand what the actual keymap is. What does "p" do, for example? Can you explain what the symbols mean?

**Reply by eugenesvk** [reply]
<kbd>p</kbd> opens the View sub-menu ☰👁
<kbd>⇧</kbd><kbd>p</kbd> opens the same View sub-menu, but with a sticky mode ☰👁ₛ (this is not yet supported by Helix, see [issue](https://github.com/helix-editor/helix/issues/1487))
<kbd>⎇</kbd><kbd>p</kbd> `ensure_selections_forward` which flips selection forward: cursor after anchor

You can hover over a button (or press the button on a keyboard with <kbd>⇧</kbd>) to see a tooltip like this that lists the command names:

<img width="455" alt="tt" src="https://user-images.githubusercontent.com/12956286/233624515-54fefcc3-c039-40e0-aba2-1963478f6253.png">

But these tooltips don't show command descriptions and they also don't work for custom command sub-menus like View since Helix doesn't support custom group names (see [issue](https://github.com/helix-editor/helix/issues/1508) and [issue](https://github.com/helix-editor/helix/issues/1752)), so for your example of <kbd>p</kbd> the tooltip is less helpful 

<img width="331" alt="tp" src="https://user-images.githubusercontent.com/12956286/233628139-900534fb-d405-476f-857d-718e36305038.png">


However, you're not supposed to magically understand all this icon soup just by looking at it, these are cheat sheets, so they are meant to support you after you've got some understanding of the config itself, so the more basic way of learning is reading the config file, it groups commands by functon, for example:
  - open the [full config](https://github.com/eugenesvk/kbdHelix/blob/d366b2382e7b08cae45c7990c917d3516dbacfc9/config/modifew.toml#L65) or the template for just the [Normal mode](https://github.com/eugenesvk/kbdHelix/blob/d366b2382e7b08cae45c7990c917d3516dbacfc9/src/m1NOR.toml.tmpl#L64) at the `#Ⓝ Move` section
  - keep the [normal mode cheat sheet](https://eugenesvk.github.io/kbdHelix/modifew/#m-normal) side-by-side
  - read the config with comments on what some commands do and then match them to the cheat sheet buttons
  - after some use you might only need the cheat sheets (with tooltips) to start associating ☰ with sub-menus and ☰👁 with a View sub-menu (that's what the eye is for)


Also you can search for `'p'` or the `☰👁` icon in the config file to see what command it maps to and what that command does

**Reply by willfaught** [reply]
If the symbols aren't meant to explain your proposal, then can you give concrete examples of your general design improvements? For instance:

>⌂ Home keys are safe: all 8 of the home keys are movement commands, so to modify a buffer your fingers need to leave ⌂!

What are the actual keys that you're referring to here (I count 11 keys on my QWERTY keyboard home row, not counting caps lock and return), and what do they do? And what do the other keys do instead? I don't want to hunt through code to figure this out, that's what posts like this are supposed to explain/summarize.

>Insert mode has a copy of most of the Normal mode commands (from key and ⇧key) on the sames keys, but with an extra ⌥ — no need to switch modes for just a single command!

I don't follow what this means concretely. Do you mean that "w" from Normal mode somehow works in Insert mode too? How does that work if "w" inserts the letter w in insert mode? What does "with an extra [option key]" mean in this context? Are you saying that when in insert mode, you can have all the same normal mode commands by pressing [option key] first and then performing them?

>Differentiate between repeatable and single-press paired keybinds (left/right, back/forward, earlier/later etc.)

Can you give an example? What scenario would be better in this case?

Edited

**Reply by eugenesvk** [reply]
> If the symbols aren't meant to explain your proposal

Of course they are, that's why 🢔ω (move left by short word) represented with a left arrow and a letter resembling __w__ in word, and big words are __W__ so it's easier not to confuse
  - ω and
  - W vs
  - w

they are just not meant to explain everything

> , then can you give concrete examples of your general design improvements? For instance:
> 
> > ⌂ Home keys are safe: all 8 of the home keys are movement commands, so to modify a buffer your fingers need to leave ⌂!
> 
> What are the actual keys that you're referring to here (I count 11 keys on my QWERTY keyboard home row, not counting caps lock and return)

You only have 8 fingers on that row, their resting positions are 8 home keys

> , and what do they do?

This is listed just 2 paragraphs below on the page (except for <kbd>s</kbd>), but the only __summary__ part relevant to this design improvement is that it's not a __destructive__ command! Otherwise it doesn't matter what the keys do, it matters for other design reasons, e.g., repeatable commands are co-located, that's why home keys with repeatable commands are listed in those other sections. But otherwise it just follows that frequent commands get to frequent keys, which explains why <kbd>s</kbd> is missing, the frequency principle has already been illustrated with other commands

>  And what do the other keys do instead?

which other keys?

>   I don't want to hunt through code to figure this out, that's what posts like this are supposed to explain/summarize.

Which it does - summarize. I mean, what you're essentially asking for is not summary, but to list all key, but then just the 8 home keys represent >20 commands if we count the modifiers and even more if we count other modes.
And that's exactly what the config file is — it's not code, but the full list of keys with commands and comments on what they do, you can read it from top to bottom instead of hunting, that's why it has comments and symbols in there and commands are grouped into sections!
The hunting comes later when you only need to check one key here and the tooltip (which lists commands, but not descriptions) is not enough. Or if you want to look up <kbd>p</kbd> instead of reading the section `#Ⓝ Minor modes / menu dialogs` which explains that View mode is mapped to <kbd>p</kbd>

(I'd like to add the descriptions to the tooltips as well as they can be more useful vs. just the command name, but these are harder to parse since this less structured data in the comments instead of key properties due to the limitations of Helix's config format, see [this issue](https://github.com/helix-editor/helix/issues/1474))


> > Insert mode has a copy of most of the Normal mode commands (from key and ⇧key) on the sames keys, but with an extra ⌥ — no need to switch modes for just a single command!
> 
> I don't follow what this means concretely. ... Are you saying that when in insert mode, you can have all the same normal mode commands by pressing [option key] first and then performing them?

So you do understand!

> > Differentiate between repeatable and single-press paired keybinds (left/right, back/forward, earlier/later etc.)
> 
> Can you give an example?

There are two sections (repeatable/single-press) with half a dozen examples. Have you checked the [full info here](https://eugenesvk.github.io/kbdHelix/modifew/) page?

>  What scenario would be better in this case?

The scenario where you use them repeatedly to go back and forth as also explained in the linked page:

> so char/word/line-based movements and in-/decrement are together rather than being on separate rows by default. This greatly simplifies going back and forth

**Comment by alexeymuranov** [comment]
A somewhat related/alternative/complementary idea: #9896.

**Comment by julien-h2** [comment]
Hey guys, great idea! I love it.

I note though, I think word-based navigation is outdated and should be deprecated. I would expect a modern keymap to provide a more modern way. 

Word are quite arbitrary. They are well adapted for plaintext navigation but not so much when programming for instance.

I don't pretend to know what it the best solution here but I have found that `jump to char` (vim f) is more precise and faster than typing word/word/word... Also `jump to char pair` is very effective for navigating accross lines. this is extremely well implemented in flash.nvim because they make it repeatable, visually highlighted, and add optional labels to jump to the n-th match immediately.

**Reply by kirawi** [reply]
In Helix, navigation can be done through the syntax tree in addition to normal movement commands. It is limited but we have issues for expanding tree-sitter commands. The [initial jump mode PR](https://github.com/helix-editor/helix/pull/5340) also included jumping to tree-sitter nodes, but I think this was deferred to the plugin system to not complicate the feature.
